plan.go

```diff
@@ -0,0 +1,172 @@
+	LogReadURL       string                `jsonapi:"attr,log-read-url"`
+	Status           PlanStatus            `jsonapi:"attr,status"`
+	StatusTimestamps *PlanStatusTimestamps `jsonapi:"attr,status-timestamps"`
+}
+
+// PlanStatusTimestamps holds the timestamps for individual plan statuses.
+type PlanStatusTimestamps struct {
+	CanceledAt   time.Time `json:"canceled-at"`
+	CreatedAt    time.Time `json:"created-at"`
+	ErroredAt    time.Time `json:"errored-at"`
+	FinishedAt   time.Time `json:"finished-at"`
+	MFAWaitingAt time.Time `json:"mfa_waiting-at"`
+	PendingAt    time.Time `json:"pending-at"`
+	QueuedAt     time.Time `json:"queued-at"`
+	RunningAt    time.Time `json:"running-at"`
+}
+
+// Read a single run by its ID.
+func (s *Plans) Read(ctx context.Context, planID string) (*Plan, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	u := fmt.Sprintf("plans/%s", url.QueryEscape(planID))
+	req, err := s.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Plan{}
+	err = s.client.do(ctx, req, p)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// Logs retrieves the logs of a plan.
+func (s *Plans) Logs(ctx context.Context, planID string) (io.Reader, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	// Get the plan to make sure it exists.
+	p, err := s.Read(ctx, planID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return an error if the log URL is empty.
+	if p.LogReadURL == "" {
+		return nil, fmt.Errorf("Plan %s does not have a log URL", planID)
+	}
+
+	return &LogReader{
+		client: s.client,
+		ctx:    ctx,
+		planID: p.ID,
+	}, nil
+}
+
+// LogReader implements io.Reader for streaming plan logs.
+type LogReader struct {
+	client *Client
+	ctx    context.Context
+	offset int64
+	planID string
+}
+
+func (r *LogReader) Read(l []byte) (int, error) {
+	for {
+		select {
+		case <-r.ctx.Done():
+			return 0, r.ctx.Err()
+		case <-time.After(500 * time.Millisecond):
+			p, err := r.client.Plans.Read(r.ctx, r.planID)
+			if err != nil {
+				return 0, err
+			}
+
+			u := fmt.Sprintf("%s?limit=%d&offset=%d", p.LogReadURL, len(l), r.offset)
+			req, err := http.NewRequest("GET", u, nil)
+			if err != nil {
+				return 0, err
+			}
+			req = req.WithContext(r.ctx)
+
+			// Retrieve the next chunk.
+			resp, err := r.client.http.Do(req)
+			if err != nil {
+				return 0, err
+			}
+
+			// Basic response checking.
+			if err := checkResponseCode(resp); err != nil {
+				resp.Body.Close()
+				return 0, err
+			}
+
+			// Check if we need to continue the loop and wait 500 miliseconds
+			// before checking if there is a new chunk available or that the
+			// plan is finished and we are done reading all chunks.
+			if resp.ContentLength == 0 {
+				resp.Body.Close()
+
+				switch p.Status {
+				case PlanCanceled, PlanErrored, PlanFinished:
+					return 0, io.EOF
+				default:
+					continue
```

Comment:
You should just be able to return `0, nil` for this case, rather than making `Read` block until more data arrives. The next `Read()` will sleep 500ms anyways. This would clear up some of the body closing logic also.